lib/collection/src/collection.rs

```diff
@@ -1520,64 +1533,77 @@ impl Collection {
+                    versioned_shard_path(&snapshot_temp_dir_path, *shard_id, 0);
                 create_dir_all(&shard_snapshot_path).await?;
                 // If node is listener, we can save whatever currently is in the storage
                 let save_wal = self.shared_storage_config.node_type != NodeType::Listener;
                 replica_set
-                    .create_snapshot(temp_dir, &shard_snapshot_path, save_wal)
+                    .create_snapshot(&snapshot_temp_dir_path, &shard_snapshot_path, save_wal)
                     .await?;
             }
         }
 
-        CollectionVersion::save(&snapshot_path_with_tmp_extension)?;
+        // Save collection config and version
+        CollectionVersion::save(&snapshot_temp_dir_path)?;
         self.collection_config
             .read()
             .await
-            .save(&snapshot_path_with_tmp_extension)?;
-
-        let snapshot_path_with_arc_extension_clone = snapshot_path_with_arc_extension.clone();
-        let snapshot_path_with_tmp_extension_clone = snapshot_path_with_tmp_extension.clone();
-        log::debug!("Archiving snapshot {:?}", snapshot_path_with_tmp_extension);
+            .save(&snapshot_temp_dir_path)?;
+
+        // Dedicated temporary file for archiving this snapshot (deleted on drop)
+        let snapshot_temp_arc_file = tempfile::Builder::new()
+            .prefix(&format!("{snapshot_name}-arc-"))
+            .tempfile_in(global_temp_dir)?;
+
+        // Archive snapshot folder into a single file
+        let snapshot_temp_arc_file_arc = Arc::new(snapshot_temp_arc_file);
+        let snapshot_arc_file_path_clone = snapshot_temp_arc_file_arc.clone();
+        let snapshot_temp_dir_path_clone = snapshot_temp_dir_path.clone();
+        log::debug!("Archiving snapshot {:?}", snapshot_temp_dir_path);
         let archiving = tokio::task::spawn_blocking(move || {
-            // have to use std here, cause TarBuilder is not async
-            let file = std::fs::File::create(&snapshot_path_with_arc_extension_clone)?;
-            let mut builder = TarBuilder::new(file);
+            let mut builder = TarBuilder::new(snapshot_arc_file_path_clone.as_ref());
             // archive recursively collection directory `snapshot_path_with_arc_extension` into `snapshot_path`
-            builder.append_dir_all(".", &snapshot_path_with_tmp_extension_clone)?;
+            builder.append_dir_all(".", &snapshot_temp_dir_path_clone)?;
             builder.finish()?;
             Ok::<_, CollectionError>(())
         });
-
         archiving.await??;
 
-        // remove temporary snapshot directory
-        remove_dir_all(&snapshot_path_with_tmp_extension).await?;
+        // Remove temporary snapshot directory (automatically dropped on previous error)
+        log::debug!(
+            "Removing temporary snapshot data {:?}",
+            snapshot_temp_dir_path
+        );
+        remove_dir_all(&snapshot_temp_dir_path).await?;
```

Comment:
We _can_ remove this because it gets unlinked automatically, right? But I guess it's fine to keep.

Reply:
You are right, it could be removed.
I like to have an explicit deletion on the happy path as soon as possible and keep the drop behavior on error. 
I can remove it if you feel strongly about it :)